### PR TITLE
[IE TESTS] Added test for NMS

### PR DESCRIPTION
**Pull request by mandrono**
Issue: 33239

### CODE REVIEW

**Comment by mandrono** [comment]
@iefode @mikhail-treskin review tests please

@vgavrilo I changed comparator for sorting in case if ```sort_result_descending == true``` to more particular comparator to get deterministic behaviour and escape situation when filtred boxes with same score have different position from launch to launch.
Take a look at this please

File: ngraph/core/reference/src/runtime/reference/non_max_suppression.cpp [hunk 1]
```diff
@@ -262,7 +262,14 @@ namespace ngraph
                 {
                     std::sort(filteredBoxes.begin(),
                               filteredBoxes.end(),
-                              [](const BoxInfo& l, const BoxInfo& r) { return l.score > r.score; });
+                              [](const BoxInfo& l, const BoxInfo& r) {
+                                  return (l.score > r.score) ||
+                                         (l.score == r.score && l.batch_index < r.batch_index) ||
+                                         (l.score == r.score && l.batch_index == r.batch_index &&
+                                          l.class_index < r.class_index) ||
```
Comment (vgavrilo) [inline ngraph/core/reference/src/runtime/reference/non_max_suppression.cpp:269]: Can you compare results of NMS-5 using such boxes comparator with results from ONNX tests?

**Review by vgavrilo — CHANGES_REQUESTED** [review]

File: ngraph/core/reference/src/runtime/reference/non_max_suppression.cpp (hunk 1 above)
Comment (mandrono) [reply ngraph/core/reference/src/runtime/reference/non_max_suppression.cpp:269]: Tests passed.
```
[ RUN      ] INTERPRETER.nonmaxsuppression_center_point_box_format
[       OK ] INTERPRETER.nonmaxsuppression_center_point_box_format (4 ms)
[ RUN      ] INTERPRETER.nonmaxsuppression_flipped_coordinates
[       OK ] INTERPRETER.nonmaxsuppression_flipped_coordinates (1 ms)
[ RUN      ] INTERPRETER.nonmaxsuppression_identical_boxes
[       OK ] INTERPRETER.nonmaxsuppression_identical_boxes (1 ms)
[ RUN      ] INTERPRETER.nonmaxsuppression_limit_output_size
[       OK ] INTERPRETER.nonmaxsuppression_limit_output_size (1 ms)
[ RUN      ] INTERPRETER.nonmaxsuppression_single_box
[       OK ] INTERPRETER.nonmaxsuppression_single_box (0 ms)
[ RUN      ] INTERPRETER.nonmaxsuppression_suppress_by_IOU
[       OK ] INTERPRETER.nonmaxsuppression_suppress_by_IOU (1 ms)
[ RUN      ] INTERPRETER.nonmaxsuppression_suppress_by_IOU_and_scores
[       OK ] INTERPRETER.nonmaxsuppression_suppress_by_IOU_and_scores (1 ms)
[ RUN      ] INTERPRETER.nonmaxsuppression_two_batches
[       OK ] INTERPRETER.nonmaxsuppression_two_batches (1 ms)
[ RUN      ] INTERPRETER.nonmaxsuppression_two_classes
[       OK ] INTERPRETER.nonmaxsuppression_two_classes (1 ms)
[----------] 9 tests from INTERPRETER (11 ms total)
```

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        IE_ASSERT(memory);
+        const auto lockedMemory = memory->wmap();
+        const auto actualBuffer = lockedMemory.as<const uint8_t *>();
+
+        if (outputIndex == 2) {
+            const int exp = *reinterpret_cast<const int *>(expectedBuffer);
+            const int act = *reinterpret_cast<const int *>(actualBuffer);
+            if (expected.size() != actual->byteSize())
+                throw std::runtime_error("Expected and actual size 3rd output have different size");
+            if (exp != act) {
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp:92]: Without threshold? And why int?

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    result << "maxOutBoxesPerClass=" << maxOutBoxesPerClass << "_";
+    result << "iouThr=" << iouThr << "_scoreThr=" << scoreThr << "_softNmsSigma=" << softNmsSigma << "_";
+    result << "boxEncoding=" << boxEncoding << "_sortResDescend=" << sortResDescend << "_outType=" << outType << "_";
+    result << "TargetDevice=" << targetDevice;
+    return result.str();
+}
+
+void NmsLayerTest::ConfigureNetwork() {
+    int i = 0;
+    for (const auto &out : cnnNetwork.getOutputsInfo()) {
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp:42]: IMO, using `for (size_t i = 0; i <cnnNetwork.getOutputsInfo().size(); i++) {` is more clear

Comment (ilyachur) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp:42]: I think it is your personal preference 😄 

`foreach` construction is easier.

File: inference-engine/src/mkldnn_plugin/nodes/non_max_suppression.cpp [hunk 4]
```diff
@@ -250,7 +250,9 @@ class NonMaxSuppressionImpl: public ExtLayerBase {
             }
             numFiltBox[batch_idx][class_idx] = fb.size();
             size_t offset = batch_idx*num_classes*max_output_boxes_per_class + class_idx*max_output_boxes_per_class;
-            cpu_memcpy(filtBoxes.data() + offset, fb.data(), fb.size() * sizeof(filteredBoxes));
+            for (size_t i = 0; i < fb.size(); i++) {
```
Comment (ilyachur) [inline inference-engine/src/mkldnn_plugin/nodes/non_max_suppression.cpp:253]: Why did you change memcpy to cycle?

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp (hunk 2 above)
Comment (mandrono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/non_max_suppression.cpp (hunk 4 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/non_max_suppression.cpp]: Implementation with ```cpu_memcpy``` work incorrect in CentOS7

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    std::tie(paramsPrec, maxBoxPrec, thrPrec) = inPrecisions;
+
+    numOfSelectedBoxes = std::min(numBoxes, maxOutBoxesPerClass) * numBatches * numClasses;
+
+    const std::vector<size_t> boxesShape{numBatches, numBoxes, 4}, scoresShape{numBatches, numClasses, numBoxes};
+    auto ngPrc = convertIE2nGraphPrc(paramsPrec);
+    auto params = builder::makeParams(ngPrc, {boxesShape, scoresShape});
+    auto paramOuts = helpers::convert2OutputVector(helpers::castOps2Nodes<op::Parameter>(params));
+
+    auto nms = std::dynamic_pointer_cast<op::v5::NonMaxSuppression>(builder::makeNms(paramOuts[0], paramOuts[1], convertIE2nGraphPrc(maxBoxPrec),
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp]: removed

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp (hunk 3 above)
Comment (iefode) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp:42]: And increasing `i` in cycle? XD So interesting)

**Review by iefode — APPROVED** [review]
In general, LGTM. Besides one comment about cycle

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp (hunk 3 above)
Comment (mandrono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/non_max_suppression.cpp]: fixed

**Review by vgavrilo — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]
